content/en/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions.md

```diff
@@ -994,6 +996,131 @@ Here is the declarations type mapping between OpenAPIv3 and CEL type:
+too many instructions, execution of the expression will be halted, and an error will result.
+Each CustomResourceDefinition is also allowed a certain amount of resources to finish executing all of 
+its CEL expressions. If the sum total of its expressions are estimated at apply time to go over that limit, 
+then a validation error will also occur.
+
+It is considered best practice to put a limit via `maxItems`, `maxProperties`, and `maxLength` for anything
+that will be processed in a CEL rule in order to prevent validation errors during cost estimation. For 
+example, given this schema with one rule:
+
+```yaml
+openAPIV3Schema:
+  type: object
+  properties:
+    foo:
+      type: array
+      items:
+          type: string
+      x-kubernetes-validations:
+        - rule: "self.all(x, x.contains('a string'))"
+```
+
+The cost system will not allow this rule. Without limits being set, the estimated cost of this rule will
+exceed the per-expression cost limit. But if we add limits in the appropriate places, the rule will be allowed:
+
+```yaml
+openAPIV3Schema:
+  type: object
+  properties:
+    foo:
+      type: array
+      maxItems: 25
+      items:
+        type: string
+        maxLength: 10
+      x-kubernetes-validations:
+        - rule: "self.all(x, x.contains('a string'))"
+```
+
+The cost estimation system takes into account how many times the rule will be executed in addition to the
+data that the rule is working with. For instance, the following rule will have the same estimated cost as the
+previous example (despite the rule now being defined on the individual array items):
+
+```yaml
+openAPIV3Schema:
+  type: object
+  properties:
+    foo:
+      type: array
+      maxItems: 25
+      items:
+        type: string
+        x-kubernetes-validations:
+          - rule: "self.contains('a string'))"
+        maxLength: 10
+```
+
+If a list inside of a list has a CEL expression that uses `self.all`, that is significantly more expensive 
+than a non-nested list with the same expression. A rule that would have been allowed on a non-nested list might need lower limits set on both nested lists in order to be allowed. Nested traversals on the same
+list (i.e, nesting calls to `self.all` or similar) should be avoided where possible for similar reasons. 
+
+In general, both systems will allow expressions that do not need to iterate; these expressions will
+always take the same amount of time regardless of how large their input is. `self.foo == 1` will be allowed.
+But if `foo` is a string and we instead have `self.foo.contains("someString")`, our expression will take 
+longer to execute depending on how long `foo` is. Another example would be if `foo` was an array, and we
+had an expression `self.foo.all(x, x > 5)`. The cost system will always assume the worst-case scenario if
+a limit on the length of `foo` is not given, and this will happen for anything that can be iterated
+over (lists, maps, etc.)
```

Comment:
This feels like a really late place in the docs to explain that cost is proportional to the amount of data traversed.  As a concept, it seems useful to know when reading much of the documentation that comes before this.